File: packages/node/src/core/auth-context.ts
/**
 * A framework-agnostic view of an inbound request, sufficient for a provider to
 * locate and validate credentials.
 *
 * Adapters (Express, Fastify, …) build an `AuthContext` from their native
 * request object. The core and providers depend only on this — never on
 * `express.Request` — which is what keeps the framework HTTP-server-agnostic.
 */
export interface AuthContext {
  /**
   * Lower-cased request headers. Multi-valued headers are preserved as arrays.
   * Providers read e.g. `headers["authorization"]` from here.
   */
  readonly headers: Readonly<Record<string, string | string[] | undefined>>;

  /**
   * The bearer token already extracted from the request, if the adapter chose to
   * do so. When present, providers should prefer it over re-parsing headers.
   */
  readonly token?: string;

  /** Request method, when known. Purely informational for providers. */
  readonly method?: string;

  /** Request path, when known. Purely informational for providers. */
  readonly path?: string;
}

/**
 * Builds a minimal {@link AuthContext} from a header map. Useful for adapters and
 * for tests that need to exercise providers without a real HTTP server.
 */
export function createAuthContext(
  headers: Record<string, string | string[] | undefined>,
  extra: { token?: string; method?: string; path?: string } = {},
): AuthContext {
  const lowerCased: Record<string, string | string[] | undefined> = {};
  for (const [key, value] of Object.entries(headers)) {
    lowerCased[key.toLowerCase()] = value;
  }
  return {
    headers: lowerCased,
    ...(extra.token !== undefined ? { token: extra.token } : {}),
    ...(extra.method !== undefined ? { method: extra.method } : {}),
    ...(extra.path !== undefined ? { path: extra.path } : {}),
  };
}

File: packages/go/core/auth_context.go
package core

import "strings"

// AuthContext is a framework-agnostic view of an inbound request, sufficient for
// a provider to locate and validate credentials. HTTP adapters build one from
// their native request; the core and providers depend only on this — never on a
// concrete server type — which keeps the framework HTTP-server-agnostic.
type AuthContext struct {
	// Headers are the request headers with lower-cased keys.
	Headers map[string][]string

	// Token is the bearer token already extracted from the request, if any.
	Token string

	// Method is the request method, when known (informational for providers).
	Method string

	// Path is the request path, when known (informational for providers).
	Path string
}

// NewAuthContext builds an AuthContext from a header map, lower-casing keys.
func NewAuthContext(headers map[string][]string, token, method, path string) AuthContext {
	lc := make(map[string][]string, len(headers))
	for k, v := range headers {
		lc[strings.ToLower(k)] = v
	}
	return AuthContext{Headers: lc, Token: token, Method: method, Path: path}
}

// Header returns the first value of the named (case-insensitive) header.
func (c AuthContext) Header(name string) string {
	if v := c.Headers[strings.ToLower(name)]; len(v) > 0 {
		return v[0]
	}
	return ""
}
